docs/design/2020-04-13-ticdc-auto-gc-safepoint-2-cn.md

```diff
@@ -0,0 +1,131 @@
+
+SetGCSafePointLimit 更新某个服务的 safepoint limit，其参数为 service_id，safepoint 以及 ttl，service_id 是用户提供的服务标识，ttl 为这个 safepoint limit 设置一个过期时间，以避免服务挂掉后，数据长期遗留下来。
+
+可以看到，SetGCSafePointLimit 的参数中，设置 service_id 为空，TTL 为 0，则实际上就是 UpdateGCSafePoint 接口的语义，即 UpdateGCSafePoint 只是 SetGCSafePointLimit 的特殊形式。
+
+**因此，本文建议为 UpdateGCSafePoint 新增 service_id，ttl 两个参数，不需要新增额外接口。**
+
+### Delete range 引起的矛盾
+
+TiDB 实际上在汇报 GC safepoint 之前，会调用 delete range 删除因 drop table 等操作而遗留的数据，delete range 成功后，TiDB 才会将 safepoint 更新到 PD。
+
+引入多个 safepoint 之后，TiDB 在 delete range 之前，不能仅依赖自身决定的 safepoint 值，还需要确保小于其他服务设置的 safepoint 值。因此，TiDB 需要先获取系统中除了自己以外的，其他服务的 safepoint 的值。
+
+从上述两节可以看到，我们消除了 limiter 的概念，因此 TiDB 通过 GetGCSafePoint 接口获取到的值，是系统最小值，这个最小值可能是 TiDB 自己维护的，这跟原来的实现有些差异：
+
+假设 TiDB 设置 `safepoint = 100`， TiCDC 设置 `safepoint = 150`，TiDB 更新下次 `safepoint = 200` 之前，执行 delete range 操作，因此需要从 PD 中获取一个合适的 safepoint，即 `safepoint = 150`。但是，如果按现在的设计，系统中的最小值是 100，则跟原来的行为逻辑不符了。这是消除 limiter 的方案带来的最大争议。
+
+这里，分三个层面解释
+
+1. 在上述情况中，TiDB 如果获取到的 `safepoint = 100`，则 100 ~ 150 之间的数据不能通过 delete range 删除掉，但在下一轮 GC 的时候，系统返回最小 `safepoint = 150`，这时 100 ~ 150 之间的数据自然会被清除掉。因此，这种情况只是推迟了数据清除的时间，不影响正确性。
+2. 稍微修改一下上述场景，假设TiCDC 设置 `safepoint = 100`， TiDB 设置 `safepoint = 150`，在 TiDB 更新下次 GC safepoint 时，从系统中获取到的 `safepoint = 100`，仍然无法 delete range 删除 100 ~ 150 之间的数据。因此， 1 只是 2 的特殊情况而已。
+3. TiDB 在汇报 GC safepoint 之前，调用 delete range 删除数据，这本身违反了 safepoint 的定义，使得大于等于 safepoint 的数据也可能被删除，是一种不严谨的实现，这个实现在只有 TiDB 的时候可能没有问题，但随着其他组件也依赖 safepoint，维护其语义的正确性也变得重要了起来。因此，强烈建议将 GC 的流程改为先更新 safepoint 再 delete range。
+
+### 兼容中心化 GC
+
+UpdateGCSafePoint 除了用来更新 GC safepoint 的值，另外一个作用是通知 TiKV 进行分布式 GC，TiKV 根据 safepoint 的值是否有变化，来决定是否执行 GC，因此，如果 TiDB 不更新这个值，则相当于没有启用分布式 GC。此时，TiDB 可以选择使用旧的 GC 方式。
+
+本文档提供的方案，第三方服务也可以调用 UpdateGCSafePoint 更新对应服务的 safepoint，如果用户在 TiDB 关闭了分布式 GC，则可能导致分布式 GC 和 TiDB 的中心化 GC 同时在执行，引起不必要的资源消耗。
+
+原有的 TiDB 中心化 GC 方案，并没有考虑第三方服务存在的情况，要想兼容这种情况，需要一种解耦的方式，将 TiDB 和第三方服务关联起来。TiDB 需要知道自己的 safepoint 不一定是全局的 safepoint，因此，TiDB 更新 safepoint 后，必须获取到全局 safepoint。获取到全局 safepoint 后， TiDB 执行两种决策：触发分布式 GC，或执行中心化 GC。
+
+因此操作过程如下：
+
+1. TiDB 更新 safepoint 并获取到全局 safepoint，调用 UpdateServiceGCSafepoint 实现
+2. TiDB 决策执行分布式 GC，调用 UpdateGCSafePoint
+3. TiKV 调用 GetGCSafePoint，判断是否需要 GC
+
+## 实现（PD）
+
+### 新增 UpdateServiceGCSafePoint 接口
+
+```
+message UpdateGCSafePointRequest {
+  RequestHeader header = 1;
+  uint64 safe_point = 2;
+  bytes service_id = 3;
+  int64 TTL = 4;
+}
+```
+
+数据保存路径为 `{prefix}/safe_point/service/{service_id}`，TTL 为数据的存活时长。
+
+### 维护 safepoint 最小值
+
+```golang
+func (s *Server) UpdateServiceGCSafePoint(ctx context.Context, request pdpb.UpdateGCSafePointRequest) (pdpb.UpdateGCSafePointResponse, error) {
+
+    ...
+
+    serviceID := string(request.ServiceId)
+    safePoint := request.SafePoint
+    ttl := request.TTL
+    minSafePoint := atomic.LoadUint64(&s.minSafePoint)
+
+    // Only save the safe point if it's greater than the previous one
+    if safePoint > minSafePoint {
```

Comment:
```suggestion
    if safePoint >= minSafePoint {
```

see https://github.com/pingcap/pd/pull/2633. Maybe keep it consistency with the current version would be better?